### PR TITLE
E2E tests for the Source IP Preservation for LoadBalancers

### PR DESCRIPTION
**Pull request by girishkalele**
Breaking out E2E changes from the main PR - these tests require the Alpha feature gate turned on for this feature otherwise they will consistently fail.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31138)

### CODE REVIEW

File: test/e2e/service.go [hunk 1]
```diff
@@ -966,6 +968,89 @@ var _ = framework.KubeDescribe("Services", func() {
 			framework.Failf("expected un-ready endpoint for Service %v within %v, stdout: %v", t.name, kubeProxyLagTimeout, stdout)
 		}
 	})
+
+	It("should be able to create services of type LoadBalancer and externalTraffic=localOnly [Slow]", func() {
```
Comment (jlowdermilk) [inline test/e2e/service.go]: Should probably add a [Feature:externalTrafficLocalOnly] tag so this doesn't get picked up by regular serial tests.

Comment (girishkalele) [inline test/e2e/service.go]: Done.

**Comment by girishkalele** [comment]
Build failure expected because this PR depends on the new annotations.go changes.

```
Error lines from build-log.txt
... skipping 217 lines ...
# k8s.io/kubernetes/test/e2e
test/e2e/service.go:990: undefined: service.AnnotationExternalTraffic
test/e2e/service.go:990: undefined: service.AnnotationValueExternalTrafficLocal
test/e2e/service.go:1006: undefined: service.GetServiceHealthCheckNodePort
make[1]: *** [all] Error 1
Makefile:79: recipe for target 'all' failed
Makefile:249: recipe for target 'cross' failed
```

**Comment by thockin** [comment]
Test LGTM.


@ixdy @fejta @lavalamp We need to decide how we test alpha features that are flag gated by default.

**Comment by jlowdermilk** [comment]
For e2e testing flag gated features, I think we should add an e2e job that explicitly targets them with `ginkgo.focus=\[Feature:..\]`, and set env var to turn appropriate flag gates on. This follows the pattern we've used for other feature testing, e.g. [autoscaling](https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml#L91). I'll look at adding/owning that job, and the future GKE one.

**Comment by girishkalele** [comment]
IIUC, this would be the KUBE_FEATURE_GATES environment variable (I see it in cluster/gce/config-default.sh) which would need to be picked up by config-test.sh also ? And it would need to be in place when we do the e2e cluster --up ?

```
# Optional: set feature gates
FEATURE_GATES="${KUBE_FEATURE_GATES:-}"
```

**Comment by jlowdermilk** [comment]
@girishkalele, yup. I sent https://github.com/kubernetes/test-infra/pull/427 to add an e2e job for testing this on a cluster with all api groups + non-api features enabled.

**Comment by lavalamp** [comment]
Random thought: we should embed in the test name something that identifies
it as needing a particular feature on?

**Comment by girishkalele** [comment]
@lavalamp 

I put in the Feature: qualifier in the It() clause, as @jlowdermilk suggested - is that whole string in the It() the name of the test ?

`It("should be able to create services of type LoadBalancer and externalTraffic=localOnly [Slow][Feature:externalTrafficLocalOnly]", func() {`

**Comment by lavalamp** [comment]
Cool, that seems right to me.

**Comment by girishkalele** [comment]
@thockin E2E tests for ESIPP - should be ready to start testing 

From @jlowdermilk https://github.com/kubernetes/test-infra/pull/427 needs LGTM - the node feature gate cmdline flags for kube-proxy fix is merged.

File: test/e2e/service.go [hunk 2]
```diff
@@ -966,6 +968,89 @@ var _ = framework.KubeDescribe("Services", func() {
 			framework.Failf("expected un-ready endpoint for Service %v within %v, stdout: %v", t.name, kubeProxyLagTimeout, stdout)
 		}
 	})
+
+	It("should be able to create services of type LoadBalancer and externalTraffic=localOnly [Slow][Feature:externalTrafficLocalOnly]", func() {
```
Comment (jlowdermilk) [inline test/e2e/service.go]: @girishkalele looks like the pattern used elsewhere is Feature:CapitalFeatureName, so this should be `[Feature:ExternalTrafficLocalOnly]`. Can you fix and update in [job spec](https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml#L165)?

**Pull request merged**